state.go

```diff
@@ -0,0 +1,398 @@
+	}
+
+	for _, token := range tokens {
+		if token == "" {
+			continue
+		}
+
+		token, err := vm.AllocateTokenAs(Token(token))
+		if err != nil {
+			return err
+		}
+
+		proxy.Lock()
+		proxy.tokenToVM[token] = &tokenInfo{
+			state: tokenStateAllocated,
+			vm:    vm,
+		}
+		proxy.Unlock()
+
+		session := vm.findSessionByToken(token)
+		if session == nil {
+			vm.Lock()
+			for token := range vm.tokenToSession {
+				_ = vm.freeTokenUnlocked(token)
+				delete(vm.tokenToSession, token)
+			}
+			vm.Unlock()
+			return fmt.Errorf("unknown token %s", token)
+		}
+
+		// Signal that the process is already started
+		close(session.processStarted)
+	}
+
+	return nil
+}
+
+// Returns (false, nil) if it's a clean start (i.e. no state was found).
+// Returns (false, error) if the restoring failed.
+// Returns (true, nil) if the restoring succeeded.
+func restoreAllState(proxy *proxy) (bool, error) {
+	if _, err := os.Stat(storeStateDir); os.IsNotExist(err) {
+		err := os.MkdirAll(storeStateDir, proxyStateDirPerm)
+		if err != nil {
+			return false, fmt.Errorf(
+				"couldn't create directory %s: %v",
+				storeStateDir, err)
+		}
+		return false, nil
+	}
+
+	proxyLog.Info("Restoring proxy's state from: ", proxyStateFilePath)
+
+	fdata, err := ioutil.ReadFile(proxyStateFilePath)
+	if err != nil {
+		return false, fmt.Errorf("couldn't read state file %s: %v",
+			proxyStateFilePath, err)
+	}
+
+	var proxyState proxyStateOnDisk
+
+	err = json.Unmarshal(fdata, &proxyState)
+	if err != nil {
+		return false, fmt.Errorf("couldn't unmarshal %s: %v",
+			proxyStateFilePath, err)
+	}
+
+	proxyLog.Debugf("proxyState: %+v", proxyState)
+
+	if len(proxyState.ContainerIDs) == 0 {
+		return false, fmt.Errorf("containerIDs list is empty")
+	}
+
+	if proxyState.Version > stateFileFormatVersion {
+		return false, fmt.Errorf("stored state format version (%d) is"+
+			" higher than supported (%d). Aborting",
+			proxyState.Version, stateFileFormatVersion)
+	}
+
+	proxy.socketPath = proxyState.SocketPath
+	proxy.enableVMConsole = proxyState.EnableVMConsole
+
+	for _, containerID := range proxyState.ContainerIDs {
+		go func(contID string) {
+			// ignore failures here but log them inside
+			_ = restoreVMState(proxy, contID)
+		}(containerID)
+	}
+
+	return true, nil
+}
+
+// On success returns nil, otherwise an error string message.
+func storeProxyState(proxy *proxy) error {
+	proxy.Lock()
+	defer proxy.Unlock()
+
+	// if there are 0 VMs then remove state from disk
+	if (len(proxy.vms)) == 0 {
+		if _, err := os.Stat(proxyStateFilePath); os.IsNotExist(err) {
+			return nil
+		}
+
+		if err := os.Remove(proxyStateFilePath); err != nil {
+			return fmt.Errorf("couldn't remove file %s: %v",
+				proxyStateFilePath, err)
+		}
+	}
+
+	proxyState := &proxyStateOnDisk{
+		Version:         stateFileFormatVersion,
+		SocketPath:      proxy.socketPath,
+		EnableVMConsole: proxy.enableVMConsole,
+		ContainerIDs:    make([]string, 0, len(proxy.vms)),
```

Comment:
You don't need to allocate dynamically the size of the buffer. Calling into `append()` will do that for you.

Reply:
Yeah, I know that. It looks like a premature optimization, but on the other hand does it really hurt?

Reply:
Well I don't consider this as an optimization since you could directly remove this line.

Reply:
If you insist, I'll remove this line, no problem. My intention was to avoid re-allocations on each append operation.